kedro/contrib/io/cached/cached.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018-2019 QuantumBlack Visual Analytics Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND
+# NONINFRINGEMENT. IN NO EVENT WILL THE LICENSOR OR OTHER CONTRIBUTORS
+# BE LIABLE FOR ANY CLAIM, DAMAGES, OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF, OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# The QuantumBlack Visual Analytics Limited (“QuantumBlack”) name and logo
+# (either separately or in combination, “QuantumBlack Trademarks”) are
+# trademarks of QuantumBlack. The License does not grant you any right or
+# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
+# Trademarks or any confusingly similar mark as a trademark for your product,
+#     or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains ``CachedDataSet``, a dataset wrapper which caches in memory the data saved,
+so that the user avoids io operations with slow storage media
+"""
+import pickle
+from typing import Any, Dict, Union
+
+from kedro.io import (
+    AbstractDataSet,
+    ExistsMixin,
+    FilepathVersionMixIn,
+    MemoryDataSet,
+    Version,
+)
+from kedro.io.core import VERSIONED_FLAG_KEY
+
+
+class CachedDataSet(AbstractDataSet, ExistsMixin, FilepathVersionMixIn):
+    """
+    ``CachedDataSet`` is a dataset wrapper which caches in memory the data saved, so that the
+    user avoids io operations with slow storage media.
+
+    You can also specify ``CachedDataSet``s in catalog.yml as follows:
+    ::
+      test_ds:
+          type: kedro.contrib.io.cached.CachedDataSet
+          versioned: true
+          dataset:
+              type: CSVLocalDataSet
+              filepath: example.csv
+
+    Please notice that if your dataset is versioned, this should be indicated in the wrapper class
+    as shown above
+    """
+
+    def __init__(
+        self,
+        dataset: Union[AbstractDataSet, Dict],
+        max_loads: int = None,
+        version: Version = None,
+    ):
+        super().__init__()
+        if isinstance(dataset, Dict):
+            self._dataset = self._from_config(dataset, version or Version(None, None))
```

Comment:
```suggestion
            self._dataset = self._from_config(dataset, version)
```
otherwise the wrapped dataset is always going to be versioned even if `CachedDataSet` is not versioned